packages/web-components/fast-foundation/src/design-token/core/design-token-node.ts

```diff
@@ -0,0 +1,694 @@
+            current = current.parent;
+        } while (current !== null);
+
+        return collected;
+    }
+
+    /**
+     * Resolves the local value for a token if it is assigned, otherwise returns undefined.
+     */
+    private static getLocalTokenValue<T>(
+        node: DesignTokenNode,
+        token: DesignToken<T>
+    ): StaticDesignTokenValue<T> | undefined {
+        return !DesignTokenNode.isAssigned(node, token)
+            ? undefined
+            : DesignTokenNode.isDerivedFor(node, token)
+            ? node._derived.get(token)!.value
+            : node._values.get(token);
+    }
+
+    private static getOrCreateDependencyGraph(
+        node: DesignTokenNode,
+        token: DesignToken<any>
+    ): Set<DerivedValue<any>> {
+        let dependents = node.dependencyGraph.get(token);
+
+        if (dependents) {
+            return dependents;
+        }
+
+        dependents = new Set<DerivedValue<any>>();
+        node.dependencyGraph.set(token, dependents);
+
+        return dependents;
+    }
+
+    /**
+     * Emit all queued notifications
+     */
+    private static notify() {
+        for (const record of this._notifications) {
+            record.notify();
+        }
+
+        this._notifications = [];
+    }
+
+    private static queueNotification(...records: DesignTokenChangeRecordImpl<any>[]) {
+        this._notifications.push(...records);
+    }
+
+    /**
+     * Retrieves all tokens assigned directly to a node.
+     * @param node - the node to retrieve assigned tokens for
+     * @returns
+     */
+    public static getAssignedTokensForNode(node: DesignTokenNode): DesignToken<any>[] {
+        return Array.from(node._values.keys());
+    }
+
+    /**
+     * Retrieves all tokens assigned to the node and ancestor nodes.
+     * @param node - the node to compose assigned tokens for
+     */
+    public static composeAssignedTokensForNode(
+        node: DesignTokenNode
+    ): DesignToken<any>[] {
+        const tokens = new Set(DesignTokenNode.getAssignedTokensForNode(node));
+        let current = node.parent;
+
+        while (current !== null) {
+            const assignedTokens = DesignTokenNode.getAssignedTokensForNode(current);
+
+            for (const token of assignedTokens) {
+                tokens.add(token);
+            }
+
+            current = current.parent;
+        }
+
+        return Array.from(tokens);
+    }
+
+    /**
+     * Tests if a token is assigned directly to a node
+     * @param node - The node to test
+     * @param token  - The token to test
+     * @returns
+     */
+    public static isAssigned<T>(node: DesignTokenNode, token: DesignToken<T>) {
+        return node._values.has(token);
+    }
+
+    /**
+     * The parent node
+     */
+    public get parent(): DesignTokenNode | null {
+        return this._parent;
+    }
+
+    public get children(): DesignTokenNode[] {
+        return Array.from(this._children);
```

Comment:
Should it return a new array every time?

Reply:
I think so, but I'm open to suggestions. The children are stored in a Set privately to make identity lookups easier, but I don't want that set mutable publicly so I convert it to an array for public consumption.